File: vtkm/filter/entity_extraction/testing/UnitTestGhostCellRemove.cxx
//============================================================================
//  Copyright (c) Kitware, Inc.
//  All rights reserved.
//  See LICENSE.txt for details.
//
//  This software is distributed WITHOUT ANY WARRANTY; without even
//  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
//  PURPOSE.  See the above copyright notice for more information.
//============================================================================

#include <vtkm/cont/DataSet.h>
#include <vtkm/cont/DataSetBuilderExplicit.h>
#include <vtkm/cont/DataSetBuilderRectilinear.h>
#include <vtkm/cont/DataSetBuilderUniform.h>
#include <vtkm/cont/testing/Testing.h>

#include <vtkm/filter/entity_extraction/GhostCellRemove.h>

namespace
{

vtkm::cont::ArrayHandle<vtkm::UInt8> StructuredGhostCellArray(vtkm::Id nx,
                                                              vtkm::Id ny,
                                                              vtkm::Id nz,
                                                              int numLayers,
                                                              bool addMidGhost = false)
{
  vtkm::Id numCells = nx * ny;
  if (nz > 0)
    numCells *= nz;

  constexpr vtkm::UInt8 normalCell = vtkm::CellClassification::Normal;
  constexpr vtkm::UInt8 duplicateCell = vtkm::CellClassification::Ghost;

  vtkm::cont::ArrayHandle<vtkm::UInt8> ghosts;
  ghosts.Allocate(numCells);
  auto portal = ghosts.WritePortal();
  for (vtkm::Id i = 0; i < numCells; i++)
  {
    if (numLayers == 0)
      portal.Set(i, normalCell);
    else
      portal.Set(i, duplicateCell);
  }

  if (numLayers > 0)
  {
    //2D case
    if (nz == 0)
    {
      for (vtkm::Id i = numLayers; i < nx - numLayers; i++)
        for (vtkm::Id j = numLayers; j < ny - numLayers; j++)
          portal.Set(j * nx + i, normalCell);
    }
    else
    {
      for (vtkm::Id i = numLayers; i < nx - numLayers; i++)
        for (vtkm::Id j = numLayers; j < ny - numLayers; j++)
          for (vtkm::Id k = numLayers; k < nz - numLayers; k++)
            portal.Set(k * nx * ny + j * nx + i, normalCell);
    }
  }

  if (addMidGhost)
  {
    if (nz == 0)
    {
      vtkm::Id mi = numLayers + (nx - numLayers) / 2;
      vtkm::Id mj = numLayers + (ny - numLayers) / 2;
      portal.Set(mj * nx + mi, duplicateCell);
    }
    else
    {
      vtkm::Id mi = numLayers + (nx - numLayers) / 2;
      vtkm::Id mj = numLayers + (ny - numLayers) / 2;
      vtkm::Id mk = numLayers + (nz - numLayers) / 2;
      portal.Set(mk * nx * ny + mj * nx + mi, duplicateCell);
    }
  }
  return ghosts;
}

vtkm::cont::DataSet MakeUniform(vtkm::Id numI,
                                vtkm::Id numJ,
                                vtkm::Id numK,
                                int numLayers,
                                std::string& ghostName,
                                bool addMidGhost = false)
{
  vtkm::cont::DataSet ds;

  if (numK == 0)
    ds = vtkm::cont::DataSetBuilderUniform::Create(vtkm::Id2(numI + 1, numJ + 1));
  else
    ds = vtkm::cont::DataSetBuilderUniform::Create(vtkm::Id3(numI + 1, numJ + 1, numK + 1));
  auto ghosts = StructuredGhostCellArray(numI, numJ, numK, numLayers, addMidGhost);

  if (ghostName == "default")
  {
    ds.SetGhostCellField(ghosts);
  }
  else
  {
    ds.SetGhostCellField(ghostName, ghosts);
  }
  return ds;
}

vtkm::cont::DataSet MakeRectilinear(vtkm::Id numI,
                                    vtkm::Id numJ,
                                    vtkm::Id numK,
                                    int numLayers,
                                    std::string& ghostName,
                                    bool addMidGhost = false)
{
  vtkm::cont::DataSet ds;
  auto nx(static_cast<std::size_t>(numI + 1));
  auto ny(static_cast<std::size_t>(numJ + 1));

  std::vector<float> x(nx), y(ny);
  for (std::size_t i = 0; i < nx; i++)
    x[i] = static_cast<float>(i);
  for (std::size_t i = 0; i < ny; i++)
    y[i] = static_cast<float>(i);

  if (numK == 0)
    ds = vtkm::cont::DataSetBuilderRectilinear::Create(x, y);
  else
  {
    auto nz(static_cast<std::size_t>(numK + 1));
    std::vector<float> z(nz);
    for (std::size_t i = 0; i < nz; i++)
      z[i] = static_cast<float>(i);
    ds = vtkm::cont::DataSetBuilderRectilinear::Create(x, y, z);
  }

  auto ghosts = StructuredGhostCellArray(numI, numJ, numK, numLayers, addMidGhost);

  if (ghostName == "default")
  {
    ds.SetGhostCellField(ghosts);
  }
  else
  {
    ds.SetGhostCellField(ghostName, ghosts);
  }

  return ds;
}

template <class CellSetType, vtkm::IdComponent NDIM>
static void MakeExplicitCells(const CellSetType& cellSet,
                              vtkm::Vec<vtkm::Id, NDIM>& dims,
                              vtkm::cont::ArrayHandle<vtkm::IdComponent>& numIndices,
                              vtkm::cont::ArrayHandle<vtkm::UInt8>& shapes,
                              vtkm::cont::ArrayHandle<vtkm::Id>& conn)
{
  using Connectivity = vtkm::internal::ConnectivityStructuredInternals<NDIM>;

  vtkm::Id nCells = cellSet.GetNumberOfCells();
  vtkm::Id connLen = (NDIM == 2 ? nCells * 4 : nCells * 8);

  conn.Allocate(connLen);
  shapes.Allocate(nCells);
  numIndices.Allocate(nCells);

  Connectivity structured;
  structured.SetPointDimensions(dims);

  vtkm::Id idx = 0;
  for (vtkm::Id i = 0; i < nCells; i++)
  {
    auto ptIds = structured.GetPointsOfCell(i);
    for (vtkm::IdComponent j = 0; j < NDIM; j++, idx++)
      conn.WritePortal().Set(idx, ptIds[j]);

    shapes.WritePortal().Set(i, (NDIM == 4 ? vtkm::CELL_SHAPE_QUAD : vtkm::CELL_SHAPE_HEXAHEDRON));
    numIndices.WritePortal().Set(i, NDIM);
  }
}

vtkm::cont::DataSet MakeExplicit(vtkm::Id numI,
                                 vtkm::Id numJ,
                                 vtkm::Id numK,
                                 int numLayers,
                                 std::string& ghostName)
{
  using CoordType = vtkm::Vec3f_32;

  vtkm::cont::DataSet dsUniform = MakeUniform(numI, numJ, numK, numLayers, ghostName);

  auto coordData = dsUniform.GetCoordinateSystem(0).GetDataAsMultiplexer();
  vtkm::Id numPts = coordData.GetNumberOfValues();
  vtkm::cont::ArrayHandle<CoordType> explCoords;

  explCoords.Allocate(numPts);
  auto explPortal = explCoords.WritePortal();
  auto cp = coordData.ReadPortal();
  for (vtkm::Id i = 0; i < numPts; i++)
    explPortal.Set(i, cp.Get(i));

  vtkm::cont::UnknownCellSet cellSet = dsUniform.GetCellSet();
  vtkm::cont::ArrayHandle<vtkm::Id> conn;
  vtkm::cont::ArrayHandle<vtkm::IdComponent> numIndices;
  vtkm::cont::ArrayHandle<vtkm::UInt8> shapes;
  vtkm::cont::DataSet ds;

  if (cellSet.IsType<vtkm::cont::CellSetStructured<2>>())
  {
    vtkm::Id2 dims(numI, numJ);
    MakeExplicitCells(
      cellSet.AsCellSet<vtkm::cont::CellSetStructured<2>>(), dims, numIndices, shapes, conn);
    ds = vtkm::cont::DataSetBuilderExplicit::Create(
      explCoords, vtkm::CellShapeTagQuad(), 4, conn, "coordinates");
  }
  else if (cellSet.IsType<vtkm::cont::CellSetStructured<3>>())
  {
    vtkm::Id3 dims(numI, numJ, numK);
    MakeExplicitCells(
      cellSet.AsCellSet<vtkm::cont::CellSetStructured<3>>(), dims, numIndices, shapes, conn);
    ds = vtkm::cont::DataSetBuilderExplicit::Create(
      explCoords, vtkm::CellShapeTagHexahedron(), 8, conn, "coordinates");
  }

  auto ghosts = StructuredGhostCellArray(numI, numJ, numK, numLayers);

  if (ghostName == "default")
  {
    ds.SetGhostCellField(ghosts);
  }
  else
  {
    ds.SetGhostCellField(ghostName, ghosts);
  }

  return ds;
}

void TestGhostCellRemove()
{
  // specify some 2d tests: {numI, numJ, numK, numGhostLayers}.
  std::vector<std::vector<vtkm::Id>> tests2D = { { 4, 4, 0, 2 },  { 5, 5, 0, 2 },  { 10, 10, 0, 3 },
                                                 { 10, 5, 0, 2 }, { 5, 10, 0, 2 }, { 20, 10, 0, 3 },
                                                 { 10, 20, 0, 3 } };
  std::vector<std::vector<vtkm::Id>> tests3D = { { 4, 4, 4, 2 },    { 5, 5, 5, 2 },
                                                 { 10, 10, 10, 3 }, { 10, 5, 10, 2 },
                                                 { 5, 10, 10, 2 },  { 20, 10, 10, 3 },
                                                 { 10, 20, 10, 3 } };

  std::vector<std::vector<vtkm::Id>> tests;

  tests.insert(tests.end(), tests2D.begin(), tests2D.end());
  tests.insert(tests.end(), tests3D.begin(), tests3D.end());

  for (auto& t : tests)
  {
    vtkm::Id nx = t[0], ny = t[1], nz = t[2];
    int nghost = static_cast<int>(t[3]);
    for (int layer = 0; layer < nghost; layer++)
    {
      std::vector<std::string> dsTypes = { "uniform", "rectilinear", "explicit" };
      for (auto& dsType : dsTypes)
      {
        std::vector<std::string> nameTypes = { "default", "user-specified" };
        for (auto& nameType : nameTypes)
        {
          vtkm::cont::DataSet ds;
          if (dsType == "uniform")
            ds = MakeUniform(nx, ny, nz, layer, nameType);
          else if (dsType == "rectilinear")
            ds = MakeRectilinear(nx, ny, nz, layer, nameType);
          else if (dsType == "explicit")
            ds = MakeExplicit(nx, ny, nz, layer, nameType);

          std::vector<std::string> removeType = { "all", "byType" };
          for (auto& rt : removeType)
          {
            vtkm::filter::entity_extraction::GhostCellRemove ghostCellRemoval;
            ghostCellRemoval.RemoveGhostField();

            if (rt == "all")
              ghostCellRemoval.RemoveAllGhost();
            else if (rt == "byType")
              ghostCellRemoval.RemoveByType(vtkm::CellClassification::Ghost);

            auto output = ghostCellRemoval.Execute(ds);
            vtkm::Id numCells = output.GetNumberOfCells();

            //Validate the output.

            vtkm::Id numCellsReq = (nx - 2 * layer) * (ny - 2 * layer);
            if (nz != 0)
              numCellsReq *= (nz - 2 * layer);

            VTKM_TEST_ASSERT(numCellsReq == numCells, "Wrong number of cells in output");
            if (dsType == "uniform" || dsType == "rectilinear")
            {
              if (nz == 0)
              {
                VTKM_TEST_ASSERT(output.GetCellSet().CanConvert<vtkm::cont::CellSetStructured<2>>(),
                                 "Wrong cell type for explicit conversion");
              }
              else if (nz > 0)
              {
                VTKM_TEST_ASSERT(output.GetCellSet().CanConvert<vtkm::cont::CellSetStructured<3>>(),
                                 "Wrong cell type for explicit conversion");
              }
            }
            else
            {
              VTKM_TEST_ASSERT(output.GetCellSet().IsType<vtkm::cont::CellSetExplicit<>>(),
                               "Wrong cell type for explicit conversion");
            }
          }

          // For structured, test the case where we have a ghost in the 'middle' of the cells.
          // This will produce an explicit cellset.
          if (dsType == "uniform" || dsType == "rectilinear")
          {
            if (dsType == "uniform")
              ds = MakeUniform(nx, ny, nz, layer, nameType, true);
            else if (dsType == "rectilinear")
              ds = MakeRectilinear(nx, ny, nz, layer, nameType, true);

            vtkm::filter::entity_extraction::GhostCellRemove ghostCellRemoval;
            ghostCellRemoval.RemoveGhostField();
            auto output = ghostCellRemoval.Execute(ds);
            VTKM_TEST_ASSERT(output.GetCellSet().IsType<vtkm::cont::CellSetExplicit<>>(),
                             "Wrong cell type for explicit conversion");
          }
        }
      }
    }
  }
}
}

int UnitTestGhostCellRemove(int argc, char* argv[])
{
  return vtkm::cont::testing::Testing::Run(TestGhostCellRemove, argc, argv);
}

File: python_bindings/testing/unit_test_ghost_cell_remove.py
##=============================================================================
##
##  The contents of this file are covered by the Viskores license. See
##  LICENSE.txt for details.
##
##  By contributing to this file, all contributors agree to the Developer
##  Certificate of Origin Version 1.1 (DCO 1.1) as stated in DCO.txt.
##
##=============================================================================

import numpy as np

import viskores
from viskores.cont import (
    DataSetBuilderExplicit,
    DataSetBuilderRectilinear,
    DataSetBuilderUniform,
)
from viskores.filter.entity_extraction import GhostCellRemove
from viskores.python_convenience import cell_set_type_name


def make_structured_ghost_cell_array(nx, ny, nz, num_layers, add_mid_ghost=False):
    shape = (ny, nx) if nz == 0 else (nz, ny, nx)
    ghosts = np.full(shape, 1 if num_layers else 0, dtype=np.uint8)

    if num_layers > 0:
        if nz == 0:
            ghosts[num_layers : ny - num_layers, num_layers : nx - num_layers] = 0
        else:
            ghosts[
                num_layers : nz - num_layers,
                num_layers : ny - num_layers,
                num_layers : nx - num_layers,
            ] = 0

    if add_mid_ghost:
        mi = num_layers + (nx - num_layers) // 2
        mj = num_layers + (ny - num_layers) // 2
        if nz == 0:
            ghosts[mj, mi] = 1
        else:
            mk = num_layers + (nz - num_layers) // 2
            ghosts[mk, mj, mi] = 1

    return ghosts.reshape(-1)


def structured_point_coordinates(nx, ny, nz):
    if nz == 0:
        return np.array(
            [(i, j, 0.0) for j in range(ny + 1) for i in range(nx + 1)],
            dtype=np.float32,
        )
    return np.array(
        [
            (i, j, k)
            for k in range(nz + 1)
            for j in range(ny + 1)
            for i in range(nx + 1)
        ],
        dtype=np.float32,
    )


def explicit_connectivity(nx, ny, nz):
    if nz == 0:
        shapes = np.full(nx * ny, viskores.CELL_SHAPE_QUAD, dtype=np.uint8)
        num_indices = np.full(nx * ny, 4, dtype=np.int32)
        connectivity = []
        for j in range(ny):
            for i in range(nx):
                p0 = (j * (nx + 1)) + i
                connectivity.extend((p0, p0 + 1, p0 + nx + 2, p0 + nx + 1))
        return shapes, num_indices, np.asarray(connectivity, dtype=np.int64)

    shapes = np.full(nx * ny * nz, viskores.CELL_SHAPE_HEXAHEDRON, dtype=np.uint8)
    num_indices = np.full(nx * ny * nz, 8, dtype=np.int32)

    def point_id(i, j, k):
        return (k * (ny + 1) * (nx + 1)) + (j * (nx + 1)) + i

    connectivity = []
    for k in range(nz):
        for j in range(ny):
            for i in range(nx):
                connectivity.extend(
                    (
                        point_id(i, j, k),
                        point_id(i + 1, j, k),
                        point_id(i + 1, j + 1, k),
                        point_id(i, j + 1, k),
                        point_id(i, j, k + 1),
                        point_id(i + 1, j, k + 1),
                        point_id(i + 1, j + 1, k + 1),
                        point_id(i, j + 1, k + 1),
                    )
                )
    return shapes, num_indices, np.asarray(connectivity, dtype=np.int64)


def make_ghost_cell_dataset(
    dataset_type, dims, ghost_layers, ghost_name="default", add_mid_ghost=False
):
    nx, ny, nz = dims
    point_dims = (nx + 1, ny + 1) if nz == 0 else (nx + 1, ny + 1, nz + 1)

    if dataset_type == "uniform":
        dataset = DataSetBuilderUniform.Create(point_dims)
    elif dataset_type == "rectilinear":
        axes = [np.arange(size, dtype=np.float32) for size in point_dims]
        dataset = DataSetBuilderRectilinear.Create(*axes)
    elif dataset_type == "explicit":
        coords = structured_point_coordinates(nx, ny, nz)
        shapes, num_indices, connectivity = explicit_connectivity(nx, ny, nz)
        dataset = DataSetBuilderExplicit.Create(coords, shapes, num_indices, connectivity)
    else:
        raise ValueError("dataset_type must be one of: uniform, rectilinear, explicit.")

    ghosts = make_structured_ghost_cell_array(nx, ny, nz, ghost_layers, add_mid_ghost)
    if ghost_name == "default":
        dataset.SetGhostCellField(ghosts)
    else:
        dataset.SetGhostCellField(ghost_name, ghosts)
    return dataset


def expected_cell_set_type(dataset_type, nz):
    if dataset_type == "explicit":
        return "explicit"
    return "structured2d" if nz == 0 else "structured3d"


def run_case(dataset_type, dims, layer, ghost_name, remove_mode):
    dataset = make_ghost_cell_dataset(dataset_type, dims, layer, ghost_name=ghost_name)
    assert dataset.HasGhostCellField()
    assert dataset.GetGhostCellField().GetData().AsNumPy().shape[0] == dataset.GetNumberOfCells()

    ghost_cell_remove = GhostCellRemove()
    ghost_cell_remove.SetRemoveGhostField(True)

    if remove_mode == "all":
        ghost_cell_remove.SetTypesToRemoveToAll()
        assert ghost_cell_remove.AreAllTypesRemoved()
    else:
        ghost_cell_remove.SetTypesToRemove(1)
        assert not ghost_cell_remove.AreAllTypesRemoved()

    output = ghost_cell_remove.Execute(dataset)

    nx, ny, nz = dims
    expected_cells = (nx - (2 * layer)) * (ny - (2 * layer))
    if nz != 0:
        expected_cells *= nz - (2 * layer)

    assert output.GetNumberOfCells() == expected_cells
    assert cell_set_type_name(output) == expected_cell_set_type(dataset_type, nz)


def run_mid_ghost_case(dataset_type, dims, layer, ghost_name):
    dataset = make_ghost_cell_dataset(
        dataset_type, dims, layer, ghost_name=ghost_name, add_mid_ghost=True
    )
    ghost_cell_remove = GhostCellRemove()
    ghost_cell_remove.SetRemoveGhostField(True)
    output = ghost_cell_remove.Execute(dataset)
    assert cell_set_type_name(output) == "explicit"


def main():
    tests_2d = [(4, 4, 0, 2), (5, 5, 0, 2), (10, 10, 0, 3), (10, 5, 0, 2), (5, 10, 0, 2)]
    tests_3d = [(4, 4, 4, 2), (5, 5, 5, 2), (10, 10, 10, 3), (10, 5, 10, 2), (5, 10, 10, 2)]

    for nx, ny, nz, nghost in tests_2d + tests_3d:
        for layer in range(nghost):
            for dataset_type in ("uniform", "rectilinear", "explicit"):
                for ghost_name in ("default", "user-specified"):
                    for remove_mode in ("all", "byType"):
                        run_case(dataset_type, (nx, ny, nz), layer, ghost_name, remove_mode)

                    if dataset_type in ("uniform", "rectilinear"):
                        run_mid_ghost_case(dataset_type, (nx, ny, nz), layer, ghost_name)


if __name__ == "__main__":
    main()
